Extend ChatGPT Prompts with User Custom Prompts

CoderCookE opened the pull request:
In the `chatgpt.vim` plugin, added functionality to extend the default prompt templates with user-defined custom prompts. This change allows the users to create their own prompts by defining the `g:chat_gpt_custom_prompts` variable. This enhances the flexibility and applicability of the ChatGPT plugin in different usage scenarios.


To add new prompts or override existing prompts:
```
let g:chat_gpt_custom_prompts = {'hi': 'do you think this is good enough?', 'fix': 'this does not work'}
command! -range -nargs=? Hi call SendHighlightedCodeToChatGPT('hi', <q-args>)
````


ToDo:
Fix the popup menu.

CoderCookE committed:
Extend ChatGPT Prompts with User Custom Prompts

In the `chatgpt.vim` plugin, added functionality to extend the default prompt templates with user-defined custom prompts. This change allows the users to create their own prompts by defining the `g:chat_gpt_custom_prompts` variable. This enhances the flexibility and applicability of the ChatGPT plugin in different usage scenarios.

MesRot commented:
Is there point in storing the prompts in Dict? You could do it in a following manner:
```vim
command! -range -nargs=? Ask call SendHighlightedCodeToChatGPT('',<q-args>)
command! -range -nargs=? Explain call SendHighlightedCodeToChatGPT('Can you explain it?', <q-args>)
command! -range -nargs=? Rewrite call SendHighlightedCodeToChatGPT('Can you rewrite it more idiomatically?', <q-args>)
...
```
and refactor the whole dictionary approach in the SendHighlightedCodeToChatGPT


I'll try to think about a way to add the custom prompts to the popup menu

CoderCookE committed:
Extend Custom Prompts in ChatGPT Plugin

Description:
This commit introduces flexibility in custom prompt management in the ChatGPT Vim plugin. The changes allow users to add their custom prompts by extending the existing prompt template list.

Changes:
- Moved `prompt_templates` to a global variable `g:prompt_templates`.
- The global variable is now extended with any user-defined prompts in the variable `g:chat_gpt_custom_prompts`.
- Modified `SendHighlightedCodeToChatGPT`, `s:ChatGPTMenuSink`, and `ChatGPTMenu` functions to use the new global prompt template list.

This enhancement improves the customizability of the plugin for the users.

CoderCookE committed:
Allow Dynamic Command Creation

This commit refactors the `plugin/chatgpt.vim` script to handle dynamic prompts in a more efficient manner. It introduces a global `g:promptKeys` variable, which holds the keys for all prompt templates, eliminating the need to repeatedly call the `keys()` function.

The commit also modifies the `s:ChatGPTMenuSink`, `s:ChatGPTMenuFilter`, and `ChatGPTMenu` functions to use `g:promptKeys` instead of locally defined `promptKeys`.

Additionally, it introduces a more flexible approach to defining commands based on these prompts. Instead of hardcoding each command, the code now uses a loop to create commands dynamically from the keys in `g:promptKeys`.

This refactoring enhances the code's clarity and maintainability, making it easier to add or modify prompt templates in the future.

MesRot commented on plugin/chatgpt.vim:
```diff
@@ -218,23 +232,10 @@ function! SendHighlightedCodeToChatGPT(ask, context)
-  if has_key(prompt_templates, a:ask)
-    let template  = "Given the following code snippet ". prompt_templates[a:ask]
+  if has_key(g:prompt_templates, a:ask)
+    let template  = "Given the following code snippet ". g:prompt_templates[a:ask]
```
I would remove the "Given the following code snippet " from the prompt as this might produce nonsensical prompts with user prompts

CoderCookE replied:
Make sense i've updated the wrapper to only include that when highlighted text is sent.

MesRot commented on plugin/chatgpt.vim:
```diff
@@ -49,6 +49,20 @@ if !exists("g:chat_gpt_split_direction")
   let g:chat_gpt_split_direction = 'horizontal'
 endif
 
+let g:prompt_templates = {
+\ 'rewrite': 'Can you rewrite it more idiomatically?',
+\ 'review': 'Can you provide a code review for?',
+\ 'document': 'Return documentation following language pattern conventions.',
+\ 'explain': 'Can you explain it?',
+\ 'test': 'Can you write a test for it?',
+\ 'fix': 'It has an error I need you to fix.'
+\}
+
+if exists('g:chat_gpt_custom_prompts')
+  call extend(g:prompt_templates, g:chat_gpt_custom_prompts)
+endif
```
I was thinking it might be good to move these closer to where the commands are defined so it is easier to follow the logic

CoderCookE replied:
Stylistically I prefer things that are global/constants to live at the top.

MesRot replied:
Yeah this is also a good point. You could think about splitting the file to multiples where you have the function definitions and other general code in the first file and the default definitions in the second one. On the other hand this might make the readability even harder when the globals are defined in another file.

Anyways whatever you decide you can mark this conversation resolved

MesRot reviewed (commented):
Real close to what I got done on my own fork (https://github.com/MesRot/vim-chatgpt) but I added some thoughts about the differences. It might  also wise to add example to README.md how to define these custom user prompts

CoderCookE committed:
Only include code snipper wrapped when highlighted text is sent

CoderCookE committed:
Add readme for creating custom prompt templates

CoderCookE committed:
Remove type casting, the method doesn't exist

The type inference happens automatically
